lambda-natchez/src/main/scala/feral/lambda/natchez/TracedLambda.scala

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2021 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package feral.lambda.natchez
+
+import cats.effect.kernel.MonadCancelThrow
+import cats.syntax.all._
+import feral.lambda.Lambda
+import natchez.EntryPoint
+import natchez.Span
+import natchez.Trace
+
+object TracedLambda {
+
+  def apply[F[_]: MonadCancelThrow, G[_], Event: HasKernel, Result](entryPoint: EntryPoint[F])(
+      lambda: Trace[G] => Lambda[G, Event, Result]
+  )(implicit lift: LiftTrace[F, G]): Lambda[F, Event, Result] =
+    viaSpan(entryPoint) { span => (event, context) =>
+      lift.run(span)(lambda(_)(event, context.mapK(lift.liftK)))
+    }
+
+  def viaSpan[F[_]: MonadCancelThrow, Event: HasKernel, Result](entryPoint: EntryPoint[F])(
+      lambda: Span[F] => Lambda[F, Event, Result]): Lambda[F, Event, Result] = {
```

Comment:
@bpholt Here's an alternative `TracedLambda` middleware, which only requires a single effect `F` and gives you a `Span[F]` instead of a `Trace[G]`.

Does this mean less boilerplate? It depends. If all you wanted all along was a `Span[F]`, then this is perfect.

But, if you actually want a `Trace` instance (e.g., to use natchez http4s middleware), then you have a couple options.
1. Hardcode the traced effect (either `Kleisli` or `IO`) and run it with the given span. This is straightforward enough, but you lose some abstraction power.
2. Re-invent `LiftTrace[F, G]` 😉

Reply:
This works pretty well! Let's run with it.

Reply:
Cool! So you prefer this one to the other in terms of `Trace[G]`? I think I'd like to keep both forms, but maybe this should be the primary :)

Reply:
I'm using it wrapped by `XRayTracedLambda`:

```scala
package feral.lambda.natchez.xray

import _root_.natchez._
import _root_.natchez.noop.NoopSpan
import _root_.natchez.xray.{XRay, XRayEnvironment}
import cats.data._
import cats.effect.std.Random
import cats.effect.{Trace => _, _}
import feral.lambda._
import feral.lambda.natchez._

object XRayTracedLambda {
  def apply[F[_]: Async, Event: HasKernel, Result](installer: Resource[Kleisli[F, Span[F], *], Lambda[Kleisli[F, Span[F], *], Event, Result]])
                                                  (implicit LT: LiftTrace[F, Kleisli[F, Span[F], *]]): Resource[F, Lambda[F, Event, Result]] =
    installer.mapK(Kleisli.applyK(new NoopSpan[F]())).flatMap(XRayTracedLambda(_))

  def apply[F[_]: Async, Event: HasKernel, Result](lambda: Span[F] => Lambda[F, Event, Result]): Resource[F, Lambda[F, Event, Result]] =
    Resource.eval(Random.scalaUtilRandom[F]).flatMap { implicit random =>
      Resource
        .eval(XRayEnvironment[F].daemonAddress)
        .flatMap {
          case Some(addr) => XRay.entryPoint[F](addr)
          case None => XRay.entryPoint[F]()
        }
        .map(TracedLambda.viaSpan(_)(lambda))
    }

  def apply[F[_]: Async, G[_], Event: HasKernel, Result](lambda: Lambda[G, Event, Result])
                                                        (implicit LT: LiftTrace[F, G]): Resource[F, Lambda[F, Event, Result]] =
    XRayTracedLambda { span => (event, context) =>
      LT.run(span) { _ => lambda(event, context.mapK(LT.liftK)) }
    }
}
```

Here's an example of a Lambda using that middleware: https://github.com/Dwolla/rabbitmq-topology-backup/blob/xray/src/main/scala/com/dwolla/rabbitmq/topology/LambdaHandler.scala#L24-L42

Reply:
Also, here's a custom CFN resource using it: https://github.com/Dwolla/postgresql-init-custom-resource/blob/feral/db/src/main/scala/com/dwolla/postgres/init/Handler.scala#L33-L52